enhancements/installer/openstack-no-fip-install.md

```diff
@@ -0,0 +1,136 @@
+status: implementable
+---
+
+# OpenStack BYO External Connectivity
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Graduation criteria
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+Numerous customers have requested the ability to run the OpenShift on OpenStack IPI installer without using Floating IP addresses.
+In the installer, we use floating IPs as a way to set up exeternal connectivity to the OpenShift cluster. Depending on how the OpenStack cluster
+is set up, the use of floating IPs may be impossible or very inconvenient. In order to support these customers, we intend to support installations
+on these types of OpenStack clusters.
+
+## Motivation
+
+### Goals
+
+- Extend the installer to support provider networks
+- Allow the installer to be run without setting up external connectivity
+
+### Non-Goals
+
+This enhancement does not intend to provide any guidance or features that go further than allowing customers to deploy without Floating IPs. Operations like attaching loadbalancers to the entrypoints, or any day 2 operations are beyond the scope of this enhancement. 
+
+## Proposal
+
+In order to implement this feature fully, the following changes must be made:
+1. Make `externalNetwork` an optional argument in the install-config when `machinesSubnet` is set
+2. Make setting  `lbFloatingIP` when no external network is passed enforced as invalid usage
+3. Make the API and ingress VIPs dynamically assigned by default, and only assign fixed IPs if they are specified in the install config
+
+### Design Details
+
+#### Make `externalNetwork` an optional argument in the install-config when `machinesSubnet` is set
+
+Implementing this can be done by changing the validations and the defaults so that a blank external network can be passed to the installer code. We would also disable all floating IP creation in the installer when this is the case.
+
+#### Disable `lbFloatingIP` when no external network is passed
+
+In the installer, the external network is used to provision floating IPs as well as the router. The router no longer gets created when using `machinesSubnet`, so the only changes that would need to be made are changes in the installers defaults and validations that allow the a blank `lbFloatingIP` to be pased. Then we will have to make a conditional block that does not attach the floating IP when `lbFloatingIP` is blank.
+
+#### Make the API and ingress VIPs dynamically assigned by default, and only assign fixed IPs if they are specified in the install config
+
+This should be handled by adding OpenStack specific code to the baremetal-runtime-config to enable it to look up ports in a cluster. This would break the circular dependency chain the installer currently has, and allow us to allocate the VIP ports dynamically, rather than assigning fixed IPs.
```

Comment:
If the ingress and API VIPs are assigned dynamically, we cannot we those IPs in any assets created by the installer like the ignition configs, or tls certs etc. How we do we solve that?

Also this talks about `baremetal-runtime-config to enable it to look up ports in a cluster`, I think it makes sense to provide a little more guidance for that component on how it can perform that action and what constraints it needs to follow when doing that. Like can it use ports from the machineNetworks
```
openshift-install explain installconfig.networking.machineNetwork
KIND:     InstallConfig
VERSION:  v1

RESOURCE: <[]object>
  MachineNetwork is the list of IP address pools for machines. This field replaces MachineCIDR, and if set MachineCIDR must be empty or match the first entry in the list. Default is 10.0.0.0/16 for all platforms other than libvirt. For libvirt, the default is 192.168.126.0/24.

FIELDS:
    cidr <Any> -required-
      CIDR is the IP block address pool for machines within the cluster.
```

or outside it etc.

Reply:
if the ips are assigned dynamically they will probably be assigned by dhcp. Which means we won't need them in the ignition config. However the dhcp and dns systems do need to be synced up. Will this be up to user?. 
If the dns system is setup correctly, then the tls certs and any of the systems that address a host by network address, should use hostnames not ip addresses.
We do expect the networkcidr to be assigned explicitely.  We will have to make sure dns registers the dynamically picked ip addresses. 
It seems to me that perhaps we might want an operator running where hosts that are being added can submit their information (ip address, hostname, etc..) which they will get when booted.

Reply:
Neutron DHCP and DNS are documented requirements for the OpenStack IPI installer. This guarantees that when ports are created in a subnet, they are assigned a port in that subnet's allocation pool dynamically using dhcp and then added to neutron dns